actors/builtin/miner/miner_actor.go

```diff
@@ -1140,7 +1158,8 @@ func processMissingPoStFaults(rt Runtime, st *State, store adt.Store, deadlines
 	builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to load failed recovery sectors")
 
 	// Unlock sector penalty for all undeclared faults.
-	penalty, err := unlockPenalty(st, store, currEpoch, append(detectedFaultSectors, failedRecoverySectors...), pledgePenaltyForSectorUndeclaredFault)
+	penaltyCalculator := computeUndeclaredFaultPenalty(epochReward, currentTotalPower)
+	penalty, err := unlockPenalty(st, store, currEpoch, append(detectedFaultSectors, failedRecoverySectors...), penaltyCalculator)
```

Comment:
As discussed, consider returning the detectedFaultSectors _and_ failedRecoverySectors from this method, and processing the penalty from the caller.